### PR TITLE
[Config Refactor] Support Recursive Merging for Engine Args

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
## Purpose
Addresses #5 on https://github.com/vllm-project/vllm-omni/issues/2887 around config parsing, and also adds some recursive merging to be more consistent with how other dicts are merged in other parts of the code.

With this PR:
- Engine args are no longer all or nothing (i.e., flat vals shouldn't be dropped if we have overlays with engine args)
- Dictionary valued keys are also merged recursively, since we previously only merged at the top level.

Also cleaned up some inline imports that were a bit redundant in the tests.

## Test Plan
New tests added for merging config cases.

## Test Result
Config tests pass.

CC @lishunyang12 @hsliuustc0106

### CODE REVIEW

**Review by hsliuustc0106 — COMMENTED** [review]
Good refactor. The recursive merge for dictionary-valued keys ensures proper deep merging, and the new tests verify that flat fields are preserved alongside engine_args.

File: vllm_omni/config/stage_config.py [hunk 1]
```diff
@@ -427,66 +427,81 @@ class DeployConfig:
...
+    base_merge_dict = {k: v for k, v in base.items() if k in _DEEP_MERGE_KEYS}
+    overlay_merge_dict = {k: v for k, v in overlay.items() if k in _DEEP_MERGE_KEYS}
+
+    # Get the merge dict; priority is base > overlay > merged sub
+    merged_subdict = _recursive_merge_subdicts(original=base_merge_dict, update=overlay_merge_dict)
+    merged_dict = {**base, **overlay, **merged_subdict}
+    return merged_dict
+
+
+def _recursive_merge_subdicts(original: dict, update: dict):
```
Comment (lishunyang12) [inline vllm_omni/config/stage_config.py]: Mutates `original` in place and also returns it. Caller misuse waiting to happen — if the same base dict is reused across overlay merges, the nested dicts get corrupted. Either deep-copy at entry or mark every call site that the base will be mutated.

File: vllm_omni/config/stage_config.py [hunk 2]
```diff
@@ -427,66 +427,81 @@ class DeployConfig:
...
     kwargs["default_sampling_params"] = stage_data.get("default_sampling_params")
-    kwargs["engine_extras"] = engine_args
+    kwargs["engine_extras"] = flat_args
     return StageDeployConfig(**kwargs)
 
 
 _DEEP_MERGE_KEYS = frozenset({"default_sampling_params", "engine_extras", "engine_args"})
 
 
 def _deep_merge_stage(base: dict, overlay: dict) -> dict:
```
Comment (lishunyang12) [inline vllm_omni/config/stage_config.py:530]: The previous warning when a deep-merge key has non-dict value on one side (`logger.warning("Deep-merge key %r has non-dict value...")`) is gone. Silent overlay clobber on type mismatch is exactly what the warning protected against. Restore it.

File: vllm_omni/config/stage_config.py [hunk 3]
```diff
@@ -427,66 +427,81 @@ class DeployConfig:
...
+        "runtime",
     }
 )
 
 # Fields on StageDeployConfig that are populated from engine_args dict
-_STAGE_DEPLOY_FIELDS = {f.name: f for f in fields(StageDeployConfig) if f.name not in _STAGE_NON_ENGINE_KEYS}
+_STAGE_DEPLOY_FIELDS = {f.name: f for f in fields(StageDeployConfig) if f.name not in _STAGE_RESERVED_KEYS}
 
 
 def _parse_stage_deploy(stage_data: dict[str, Any]) -> StageDeployConfig:
```
Comment (lishunyang12) [inline vllm_omni/config/stage_config.py:494]: `runtime` is extracted only inside the `if "engine_args" in stage_data:` branch; the else branch falls back to `stage_data.get("devices", "0")` only. Devices resolution diverges between the two schemas — consolidate.

File: tests/test_config_factory.py [hunk 4]
```diff
@@ -812,6 +804,97 @@ def test_merge_pipeline_deploy(self):
...
+        assert len(deploy.stages) == 1
+        stage = deploy.stages[0]
+        # Check that the engine args are set
+        assert stage.tensor_parallel_size == 4
+        assert stage.enforce_eager is True
+        # Check that the other settings are also preserved
+        assert stage.gpu_memory_utilization == 0.7
+        assert stage.max_num_seqs == 8
+
+    def test_engine_extras_deep_merges_dicts_simple(self):
```
Comment (lishunyang12) [inline tests/test_config_factory.py:1001]: `"foo"` is not a real engine field — fine for testing the merge path, but doesn't cover the realistic case of a promoted dict-valued field like `compilation_config`. Add at least one test using a real `_STAGE_DEPLOY_FIELDS` key.

File: vllm_omni/config/stage_config.py [hunk 5]
```diff
@@ -427,66 +427,81 @@ class DeployConfig:
...
+
+
+def _recursive_merge_subdicts(original: dict, update: dict):
+    """Recursively merge dict valued keys in place."""
+    for k, update_v in update.items():
+        # If we have nested dict values, continue merging
+        orig_v = original.get(k)
+        if isinstance(orig_v, dict) and isinstance(update_v, dict):
+            _recursive_merge_subdicts(orig_v, update_v)
+        else:
```
Comment (hsliuustc0106) [inline vllm_omni/config/stage_config.py]: **Consider re-adding the type-mismatch warning from the old `_deep_merge_stage`.** The previous code logged a warning when a deep-merge key had a non-dict value on one side (e.g. someone writes `engine_extras: 42` instead of `engine_extras: {…}`). `_recursive_merge_subdicts` silently overwrites without warning, which means YAML type mistakes could go unnoticed.

File: vllm_omni/config/stage_config.py [hunk 6]
```diff
@@ -427,66 +427,81 @@ class DeployConfig:
...
-        engine_args = dict(stage_data["engine_args"])
-        devices = stage_data.get("runtime", {}).get("devices", stage_data.get("devices", "0"))
+        for k, v in stage_data["engine_args"].items():
+            existing = flat_args.get(k)
+            # If we have multiple dictionaries, merge recursively.
+            if isinstance(v, dict) and isinstance(existing, dict):
+                flat_args[k] = _recursive_merge_subdicts(existing, v)
+            else:
+                flat_args[k] = v
+        runtime = stage_data.get("runtime", {})
```
Comment (hsliuustc0106) [inline vllm_omni/config/stage_config.py]: **`_recursive_merge_subdicts` mutates its `original` argument in place.** Here `existing = flat_args.get(k)` may reference a dict from the original `stage_data` (via the comprehension on line 450), meaning the input config dict gets mutated as a side effect. In practice this is safe since config dicts are thrown away after parsing, but worth being aware of / documenting if intentional.

File: tests/test_config_factory.py [hunk 7]
```diff
@@ -4,11 +4,15 @@
 Unit tests for StageConfigFactory and related classes.
 """
 
+import warnings
 from dataclasses import dataclass
+from dataclasses import fields as fields
```
Comment (hsliuustc0106) [inline tests/test_config_factory.py]: **Nit: self-alias.** `from dataclasses import fields as fields` is equivalent to `from dataclasses import fields`. Harmless but unusual -- consider simplifying.

**Review by hsliuustc0106 — COMMENTED** [review]
A few non-blocking observations from reviewing the diff.

File: vllm_omni/config/stage_config.py (hunk 1 above)
Comment (alex-jw-brooks) [reply vllm_omni/config/stage_config.py]: Deep copy can be pretty expensive, so we should be try not to use it where possible. Updated the implementation to use recursive shallow copying instead and added some more tests to ensure the original isn't mutated

File: vllm_omni/config/stage_config.py (hunk 5 above)
Comment (alex-jw-brooks) [reply vllm_omni/config/stage_config.py]: Added the warning back and added another test to make sure this case resolves to what we expect

File: vllm_omni/config/stage_config.py (hunk 2 above)
Comment (alex-jw-brooks) [reply vllm_omni/config/stage_config.py:530]: Good catch - added it back

File: vllm_omni/config/stage_config.py (hunk 6 above)
Comment (alex-jw-brooks) [reply vllm_omni/config/stage_config.py]: I've updated the implementation to recursively shallow copy and return the result instead of mutating in place 🙂

File: vllm_omni/config/stage_config.py (hunk 3 above)
Comment (alex-jw-brooks) [reply vllm_omni/config/stage_config.py:494]: This isn't really a change on this PR, it's based on [this](https://github.com/vllm-project/vllm-omni/blob/main/vllm_omni/config/stage_config.py#L470)). I assume this is not on purpose?

I went ahead and consolidated them, since I agree it makes more sense

File: tests/test_config_factory.py (hunk 4 above)
Comment (alex-jw-brooks) [reply tests/test_config_factory.py:1001]: Sure, added a more complex parsing/merging test for compilation config

**Comment by Gaohan123** [comment]
Please resolve docs problem. Thanks

File: tests/test_config_factory.py [hunk 8]
```diff
@@ -1476,7 +1522,7 @@ def test_typed_kwarg_overrides_yaml(self):
     def test_none_value_skipped_yaml_wins(self):
         stages = self._stages({"max_num_seqs": None})
         assert stages[2].runtime_overrides.get("max_num_seqs") is None
-        assert "max_num_seqs" not in stages[2].yaml_engine_args
+        assert stages[2].yaml_engine_args.get("max_num_seqs") == 64
```
Comment (alex-jw-brooks) [inline tests/test_config_factory.py:1561]: Change is unrelated, but this test is failing on main - max num seqs in the config used is 64, which is where this came from.


(this file did not have a `pytestmark`)

**Review by hsliuustc0106 — COMMENTED** [review]
## Review: [Config Refactor] Support Recursive Merging for Engine Args (#3009)

### General

2 files, 238 additions, 175 deletions. Well-sized refactor with solid test coverage. All checks pass, BLOCKED (infra only).

---

### What Changed

1. **`_parse_stage_deploy()` — no longer all-or-nothing**: Previously, if `engine_args` was present in a deploy YAML, flat-level keys (like `gpu_memory_utilization`, `max_num_seqs`) were silently dropped. Now, `flat_args` is always collected from non-reserved keys, and `engine_args` merges on top. Dict-valued keys merge recursively.

2. **`_deep_merge_stage()` — truly recursive**: Previously used `{**base_val, **v}` (shallow one-level merge). Now uses a new `_get_recursively_merged_dict()` helper that recurses into nested dicts.

3. **`_get_recursively_merged_dict()`**: New helper that recursively merges dicts with `update` taking precedence. Logs a warning on type mismatches (dict vs non-dict at the same key).

4. **`_STAGE_NON_ENGINE_KEYS` → `_STAGE_RESERVED_KEYS`**: Rename + adds `"engine_args"` and `"runtime"` to the set.

5. **Import cleanup in tests**: Moved inline imports to top of file, removed ~35 redundant inline imports, added `pytestmark` markers.

---

### Test Coverage

8 new tests covering:
- Flat fields preserved when engine_args present
- Recursive merge of nested compilation_config
- Top-level and nested dict key merging
- Type mismatch handling (dict vs scalar)
- Immutability of input dicts after merge
- Scalar precedence (engine_args wins over flat)

---

### Issues

**None blocking.** A few notes:

**1. `_get_recursively_merged_dict` warning triggers only when `orig_v is not None`**

If a base key maps to `None` and the overlay provides a dict, no warning is emitted — the overlay silently replaces it. This seems intentional (None = "not set"), but worth noting since a YAML typo setting a key to `null` wouldn't warn.

**2. Rename `_STAGE_NON_ENGINE_KEYS` → `_STAGE_RESERVED_KEYS` is internal-only**

This private symbol isn't referenced outside `stage_config.py`, so the rename is safe. Confirmed.

**3. `test_none_value_skipped_yaml_wins` assertion changed**

From `assert "max_num_seqs" not in stages[2].yaml_engine_args` to `assert stages[2].yaml_engine_args.get("max_num_seqs") == 64`. This reflects the new behavior where all dataclass fields now appear in `yaml_engine_args`, not just "extras" — correct and tested.

---

### Verdict

Clean refactor that fixes the all-or-nothing engine_args bug and adds proper recursive merging. Tests are thorough and cover edge cases. Ready to merge.

File: tests/test_config_factory.py [hunk 9]
```diff
@@ -1288,17 +1378,14 @@ class TestBaseConfigInheritance:
...
-
         ci_path = Path(get_deploy_config_path("ci/qwen3_omni_moe.yaml"))
         if not ci_path.exists():
             pytest.skip("CI deploy config not found")
 
         deploy = load_deploy_config(ci_path)
         assert len(deploy.stages) == 3
         # CI overrides
-        assert deploy.stages[0].load_format == "dummy"
+        assert deploy.stages[0].load_format is None
```
Comment (alex-jw-brooks) [inline tests/test_config_factory.py:1388]: This is also unrelated and seems to be failing on `main`, but not running since we are missing the pytest mark. I don't think we are setting the load format to dummy in the overlay for this one

**Comment by alex-jw-brooks** [comment]
@Gaohan123 @lishunyang12 @hsliuustc0106 is it possible to merge this one soon? There are a couple of failing tests on `main` that aren't running because we are missing the pytest marker.
```
FAILED tests/test_config_factory.py::TestBaseConfigInheritance::test_ci_inherits_from_main - AssertionError: assert None == 'dummy'
FAILED tests/test_config_factory.py::TestSentinelDefaultPrecedence::test_none_value_skipped_yaml_wins - AssertionError: assert 'max_num_seqs' not in {'async_chunk': True, 'async_sc...
================== 2 failed, 104 passed, 18 warnings in 0.86s ===================
```

I went ahead and fixed those cases as part of this PR since they were just small issues in the tests, but would be ideal to run these tests in CI builds

**Comment by hsliuustc0106** [comment]
> @Gaohan123 @lishunyang12 @hsliuustc0106 is it possible to merge this one soon? There are a couple of failing tests on `main` that aren't running because we are missing the pytest marker.
> 
> ```
> FAILED tests/test_config_factory.py::TestBaseConfigInheritance::test_ci_inherits_from_main - AssertionError: assert None == 'dummy'
> FAILED tests/test_config_factory.py::TestSentinelDefaultPrecedence::test_none_value_skipped_yaml_wins - AssertionError: assert 'max_num_seqs' not in {'async_chunk': True, 'async_sc...
> ================== 2 failed, 104 passed, 18 warnings in 0.86s ===================
> ```
> 
> I went ahead and fixed those cases as part of this PR since they were just small issues in the tests, but would be ideal to run these tests in CI builds

let me run the CI again with 0.21.0

**Comment by alex-jw-brooks** [comment]
Thank you @hsliuustc0106!
